### PR TITLE
Major changes for better Android Studio support

### PR DESCRIPTION
**Pull request by theimpulson**
## Summary

This MR does a major restructure in the way Etar is currently present. The goal is to follow the standard layout of the modern apps generated via Android Studio to work better with it. The biggest benefit from this major restructure is that this unbreaks/restores the ability of Android Studio to analyze the files and provide suggestions while writing the code itself which is currently absent.

This MR introduces the following changes:

- Restructures Etar for better Android Studio support,
- Imports AAR generated from the external projects/modules and adds tasks to dynamically generate them,
- Drops Eclipse support,
- Drops non-required `.iml/` file.

After this MR is merged, Custom ROMs will need to move their `Android.mk`/`Android.bp` to reflect the new path to continue using Etar.

## Test

With this MR, notice that:

- Android Studio now analyzes the file on opening and provides suggestions and shows warnings,
- Builds compile without any issues

### CODE REVIEW

**Comment by theimpulson** [comment]
@Gitsaibot please review when possible. Checks have failed as now Gradle is picking up the unit tests and trying to compile them which fails. Before the tests were not even being compiled as I can see in old workflows.

The current error can be easily fixed by adding Junit dependency and runner but tests will still fail as they are accessing resources and classes which have been deprecated for quite a long time. I would prefer to fix them in another dedicated MR.

Example workflow: https://github.com/Etar-Group/Etar-Calendar/pull/1187/checks

```gradle
> Task :compileDebugUnitTestJavaWithJavac NO-SOURCE
```

**Comment by theimpulson** [comment]
@Gitsaibot conflicts resolved. Please review again.

**Review by jspricke — CHANGES_REQUESTED** [review]
I'm not sure here. From a quick look these seem problematic:

- we don't want binary files (the aar).
- Moving everything to app/ could break other use cases.
- Don't move the .gitignore.

**Comment by theimpulson** [comment]
> I'm not sure here. From a quick look these seem problematic:
> 
>     * we don't want binary files (the aar).
> 
>     * Moving everything to app/ could break other use cases.
> 
>     * Don't move the .gitignore.

Thanks for reviewing. Splitting my replies to the feedback as per the points raised.

>     * we don't want binary files (the aar).

The binaries files have been generated via Android Studio with the added tasks. I can remove them but I included them as they will allow removing compilation of other external modules from the workflow decreasing build times of the module.

I preferred having AAR as that lets Android Studio pick up the class and other stuff to stop lint from complaining about numerous errors about missing classes being compiled from those external modules.

Right now, once a Gradle sync is completed, Studio will run the task to generate the AAR automatically. 

>     * Moving everything to app/ could break other use cases.

Could you define some example use cases this would break? I can only think of compilation with the AOSP build system but that won't be an issue as I have mentioned before after moving the said makefile/blueprint file to the new path.

>     * Don't move the .gitignore.

I haven't moved it. I added a new one in the `app` module which I believe you are referring to.

**Comment by jspricke** [comment]
>I can remove them

Please do. Generated files don't belong into the version control system. Also they are frowned upon in open source projects and would make Etar unfit for F-Droid, for example.

>compilation with the AOSP

That, yes. Did you test this?
@chirayudesai any comments?

> added a new one

Why not keep everything in the toplevel one?
Maybe clean that one up as well?

- Please either fix or disable the tests.
- Please rebase and don't push merge commits (esp. merges of master).

**Comment by chirayudesai** [comment]
> AOSP compilation

It'd be a one-time change, and we can just do a simply mv downstream when we end up pulling this.

@theimpulson you should do a quick test though to make sure that's all.

**Comment by theimpulson** [comment]
>     * Please do the file move in an extra commit without changing the files. This makes it easier to follow the changes.
Done

>     * Please move the changes to the .gitignore into one commit (currently you are adding a .gitignore first and then delete it again).
Done

>     * "MUST BE RUN MANUALLY ON CLI COMPILATION" - can this be automated?
I tried a bit but Gradle wasn't picking it up for some reason. So, I skipped it for now and added an additional step instead.

>     * nit: please don't start your commit comments with "Etar: " this git is all about Etar.
Done

File: .github/workflows/android.yml [hunk 1]
```diff
@@ -33,7 +36,9 @@ jobs:
       - name: Validate Gradle Wrapper
         uses: gradle/wrapper-validation-action@v1
       - name: Build with Gradle
-        run: ./gradlew build
+        run: |
```
Comment (jspricke) [inline .github/workflows/android.yml:39]: this should be in the AAR commit.

**Review by jspricke — CHANGES_REQUESTED** [review]

File: build.gradle [hunk 2]
```diff
@@ -1,8 +1,8 @@
 plugins {
-    id 'com.android.application' version '7.2.1' apply true
```
[the remark this answers is not in the document]
Comment (theimpulson) [reply build.gradle:2]: Those plugins were active/available globally but that doesn't need to be the case considering only the `:app` module uses it now, so they have been disabled and enabled within the said module. This is how modern apps works with plugins in the standard template.

File: app/build.gradle [hunk 3]
```diff
@@ -0,0 +1,111 @@
+plugins {
```
[the remark this answers is not in the document]
Comment (theimpulson) [reply app/build.gradle:1]: Just `plugins` and `sourceSets` as you say.

File: .github/workflows/android.yml (hunk 1 above)
Comment (theimpulson) [reply .github/workflows/android.yml:39]: Done

File: build.gradle [hunk 4]
```diff
@@ -3,8 +3,14 @@ plugins {
...
     id 'org.jetbrains.kotlin.android' version '1.7.0' apply false
     id 'org.ec4j.editorconfig' version "0.0.3" apply false
+    id "org.jetbrains.gradle.plugin.idea-ext" version "1.1.5" apply true
 }
 
+// External project configuration start
+
+// Configure and automatically generate AAR of the below mentioned external
+// projects using a custom task  in order to let Android Studio work properly.
+// MUST BE RUN MANUALLY ON CLI COMPILATION
```
[the remark this answers is not in the document]
Comment (theimpulson) [reply build.gradle:13]: Done

File: tests/Android.mk [hunk 5]
```diff
@@ -1,22 +0,0 @@
-LOCAL_PATH:= $(call my-dir)
```
[the remark this answers is not in the document]
Comment (theimpulson) [reply tests/Android.mk]: Restored, this was supposed to be handled in next MR where I aim to fix tests

File: app/proguard-rules.pro [hunk 6]
```diff
@@ -1,3 +1,25 @@
+# Add project specific ProGuard rules here.
```
[the remark this answers is not in the document]
Comment (theimpulson) [reply app/proguard-rules.pro:1]: Done

File: build.gradle (hunk 2 above)
[the remark this answers is not in the document]
Comment (jspricke) [reply build.gradle:2]: can you do that in an extra commit and add your comment from above in the commit log?

File: app/build.gradle (hunk 3 above)
[the remark this answers is not in the document]
Comment (jspricke) [reply app/build.gradle:1]: same here, please do those changes in an extra commit and add an explanation in the log.

File: app/proguard.flags [hunk 7]
```diff
@@ -0,0 +1,45 @@
+-keep class com.android.calendar.selectcalendars.SelectCalendarsSyncFragment
```
Comment (jspricke) [inline app/proguard.flags:1]: you probably want to remove the `/proguard.flags` as well.

Comment (jspricke) [reply app/proguard.flags:1]: ..and rename the file here as well.

Comment (theimpulson) [reply app/proguard.flags]: Done

File: app/build.gradle (hunk 3 above)
[the remark this answers is not in the document]
Comment (theimpulson) [reply app/build.gradle:2]: Done for plugins, left `sourceSets` in the original commit as that adapts for the new file path which is what the original commit it is in intends to do.

File: build.gradle (hunk 2 above)
[the remark this answers is not in the document]
Comment (theimpulson) [reply build.gradle:2]: Done

File: build.gradle [hunk 8]
```diff
@@ -1,8 +1,8 @@
 plugins {
```
Comment (jspricke) [inline build.gradle:1]: are those enabled by default or could we simply drop this?

**Review by jspricke — COMMENTED** [review]
looking great, just one minor thing :).
Btw. did you look into the tests? Would it be worth to enable them in CI?

**Comment by jspricke** [comment]
also, did you test this with an AOSP/LineagOS build?

File: build.gradle (hunk 8 above)
Comment (theimpulson) [reply build.gradle:1]: They are required and cannot be dropped

**Comment by theimpulson** [comment]
> looking great, just one minor thing :). Btw. did you look into the tests? Would it be worth to enable them in CI?

I did look into them. They will need quite a bit of time to get fixed and running again considering they are heavily outdated. Testing is always worth the pain for ease in development later.

**Comment by theimpulson** [comment]
> also, did you test this with an AOSP/LineagOS build?

Tested with CalyxOS (Android 12L), compiles fine after updating the paths. Pushed to LineageOS Gerrit: https://review.lineageos.org/c/LineageOS/android_packages_apps_Etar/+/334280

**Review by jspricke — APPROVED** [review]
Thanks a lot for all your work and applying all my requests :).

**Pull request merged**